libkbfs/test_stallers.go

```diff
@@ -0,0 +1,288 @@
+// stalledOp, is stalled. Caller should use the returned newCtx for subsequent
+// operations for the stall to be effective. onStalled is a channel to notify
+// the caller when the stall has happened. unstall is a channel for caller to
+// unstall an Op.
+func StallMDOp(ctx context.Context, config Config, stalledOp StallableMDOp) (
+	onStalled <-chan struct{}, unstall chan<- struct{}, newCtx context.Context) {
+	onStalledCh := make(chan struct{}, 1)
+	unstallCh := make(chan struct{})
+	stallKey := newStallKey()
+	config.SetMDOps(&stallingMDOps{
+		stallOpName: stalledOp,
+		stallKey:    stallKey,
+		staller: staller{
+			stalled: onStalledCh,
+			unstall: unstallCh,
+		},
+		delegate: config.MDOps(),
+	})
+	newCtx = context.WithValue(ctx, stallKey, true)
+	return onStalledCh, unstallCh, newCtx
+}
+
+var stallerCounter uint64
+
+func newStallKey() string {
+	stallerCounter++
+	return fmt.Sprintf("stallKey-%d", stallerCounter)
+}
+
+// staller is a pair of channels. Whenever something is to be
+// stalled, a value is sent on stalled (if not blocked), and then
+// unstall is waited on.
+type staller struct {
+	stalled chan<- struct{}
+	unstall <-chan struct{}
+}
+
+func maybeStall(ctx context.Context, opName stallableOp,
+	stallOpName stallableOp, stallKey string,
+	staller staller) {
+	if opName != stallOpName {
+		return
+	}
+
+	v, ok := ctx.Value(stallKey).(bool)
+	if !ok || !v {
+		return
+	}
+
+	select {
+	case staller.stalled <- struct{}{}:
+	default:
+	}
+	<-staller.unstall
+}
+
+// stallingBlockOps is an implementation of BlockOps whose operations
+// sometimes stall. In particular, if the operation name matches
+// stallOpName, and ctx.Value(stallKey) is a key in the corresponding
+// staller is used to stall the operation.
+type stallingBlockOps struct {
+	stallOpName StallableBlockOp
+	// stallKey is a key for switching on/off stalling. If it's present in ctx,
+	// and equal to `true`, the operation is stalled. This allows us to use the
+	// ctx to control stallings
+	stallKey string
+	staller  staller
+	delegate BlockOps
+}
+
+var _ BlockOps = (*stallingBlockOps)(nil)
+
+func (f *stallingBlockOps) maybeStall(ctx context.Context, opName StallableBlockOp) {
+	maybeStall(ctx, stallableOp(opName), stallableOp(f.stallOpName),
+		f.stallKey, f.staller)
+}
+
+func (f *stallingBlockOps) Get(
+	ctx context.Context, md *RootMetadata, blockPtr BlockPointer,
+	block Block) error {
+	f.maybeStall(ctx, StallableBlockGet)
+	return f.delegate.Get(ctx, md, blockPtr, block)
+}
+
+func (f *stallingBlockOps) Ready(
+	ctx context.Context, md *RootMetadata, block Block) (
+	id BlockID, plainSize int, readyBlockData ReadyBlockData, err error) {
+	f.maybeStall(ctx, StallableBlockReady)
+	return f.delegate.Ready(ctx, md, block)
+}
+
+func (f *stallingBlockOps) Put(
+	ctx context.Context, md *RootMetadata, blockPtr BlockPointer,
+	readyBlockData ReadyBlockData) error {
+	f.maybeStall(ctx, StallableBlockPut)
+	select {
```

Comment:
It's possible there are some tests that rely on context checking and some that don't, for different methods.  We'd have to clean that stuff up as well.  I don't think it's worth doing for this PR.